README.md
# elasticsearch-curator [![Cookbook Version](http://img.shields.io/cookbook/v/elasticsearch-curator.svg)](https://community.opscode.com/cookbooks/elasticsearch-curator) [![Build Status](https://travis-ci.org/cyberflow/chef-elasticsearch-curator.svg?branch=master)](https://travis-ci.org/cyberflow/chef-elasticsearch-curator)

Chef cookbook to install and configure [elasticsearch-curator](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/index.html) from version 4.

## Tested Platforms

* ubuntu 16.04
* centos 7

## Usage

This cookbook can be used by including `elasticsearch-curator::default` in your run list and settings attributes as needed.  Alternatively, you can use the custom resources directly.

If you are using elasticsearch-curator < 5 you might want to use the cookbook version v0.2.8.

### Attributes

| Key                                  | Type   | Description                                           | Default                                                                                                                                                             |
|--------------------------------------|--------|-------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| node['elasticsearch-curator']['version']          | String | Version of elasticsearch-curator to install, nil = latest | nil |
| node['elasticsearch-curator']['repository_url'] | String | URL of elasticsearch-curator package repository | 'http://packages.elastic.co/curator/4/debian' |
| node['elasticsearch-curator']['repository_key']           | String   | elasticsearch-curator repository key | 'https://packages.elastic.co/GPG-KEY-elasticsearch' |
| node['elasticsearch-curator']['bin_path'] | String | bin path for elasticsearch-curator | '/usr/bin' |
|node['elasticsearch-curator']['username']|String|user for running curator|'curator'|
|node['elasticsearch-curator']['config_file_path']|String|path to direct curator config file|"/home/#{node['elasticsearch-curator']['username']}/.curator"|
|node['elasticsearch-curator']['action_file_path']|String|path to direct action config file|"/home/#{node['elasticsearch-curator']['username']}/.curator"|
|node['elasticsearch-curator']['cron_minute']|String|Minute to run the curator cron job|'0'|
|node['elasticsearch-curator']['cron_hour']|String|Hour to run the curator cron job|'*'|
|node['elasticsearch-curator']['config']|Hash|config elasticsearch-curator| {<br>&nbsp;&nbsp;'client' => {<br>&nbsp;&nbsp;&nbsp;&nbsp;'hosts' => ['127.0.0.1'],<br>&nbsp;&nbsp;&nbsp;&nbsp;'port' => 9200,<br>&nbsp;&nbsp;&nbsp;&nbsp;'use_ssl' => false,<br>&nbsp;&nbsp;&nbsp;&nbsp;'ssl_no_validate' => false,<br>&nbsp;&nbsp;&nbsp;&nbsp;'timeout' => 30,<br>&nbsp;&nbsp;&nbsp;&nbsp;'master_only' => false<br>&nbsp;&nbsp;},<br>&nbsp;&nbsp;'logging' => {<br>&nbsp;&nbsp;&nbsp;&nbsp;'loglevel' => 'INFO',<br>&nbsp;&nbsp;&nbsp;&nbsp;'logformat' => 'default'<br>&nbsp;&nbsp;}<br>}|

This cookbook ships with custom resources for install elasticsearch-curator and managing the configuration file:

### Custom Resources

#### elasticsearch_curator_install

Installs elasticsearch-curator. Optionally specifies a version, otherwise the latest available is installed

```ruby
elasticsearch_curator_install 'curator' do
  install_method node['elasticsearch-curator']['install_method']
  action :install
end
```

#### elasticsearch_curator_config

Writes out the elasticsearch-curator configuration file.

```ruby
elasticsearch_curator_config 'default' do
  config node['elasticsearch-curator']['config']
  action :configure
end
```
This method also supports a http_auth property to allow passing a string with this format : "username:password". This allows retrieving the credentials from the wrapper cookbook (for example using chef-vault) and not store this sensitive information in the attributes.

#### elasticsearch_curator_action

This will setup a cron job and create action.yaml file for elasticsearch-curator.

```ruby
elasticsearch_curator_action 'action' do
  config node['elasticsearch-curator']['action_config']
  minute '0'
  hour '*'
  action :create
end
```

## Tests

To run tests, install all dependencies with [bundler](http://bundler.io/):

    bundle install
    bundle exec cookstyle
    bundle exec foodcritic .


CHANGELOG.md
# CHANGELOG
## 0.3.0
* Add support curator v5 (contributed by @EmFl)

## 0.2.8
* Fix user permissions for curator config file (contributed by @codylewandowski)
* Change serverspec tests to inspec
* Add inspec verifier to kitchen.yml

## 0.2.7
* Sets tricter permissions on the curator config (contributed by @kovyrin)

## 0.2.6
* fix issue #10

## 0.2.5
* Added support for centos 6 (contributed by @mdarii)

* Change syntax test to `cookstyle`

* Added LICENSE file

## 0.2.4
* Set conditional repo_url attributes directly in default attributes files instead of inside the resource as @Cyberflow recommended (contributed by @GHostQC)

* Fixed coding style according to Rubocop tests as @Cyberflow recommanded (contributed by @GHostQC)

* Set max blocklength to 50 instead 30 (contributed by @GHostQC)

## 0.2.3
* Fixed pip-installation on both ubuntu and centos boxes, ServerSpec test now verify successfully (contributed by @GHostQC)

## 0.2.2
* Added ServerSpec test suites (contributed by @GHostQC)

## 0.2.1
* Patched resource config.rb attributes http_auth to allow nil as a value as it is now deprecated (contributed by @GHostQC)

## 0.2.0
* Added support for rhel platform family compatible with YUM; Tested on CentOS 7.2 (contributed by @GHostQC)

## 0.1.5
* Change package name for elasticsearch-curator debian package version from 4.2

## 0.1.4
* Supports a http_auth property (contributed by @EmFl)

## 0.1.3
* Add elasticsearch-curator bin path to resource and attributes

## 0.1.2
* Bug fixed
* Add test cookbook for kitchen tests

## 0.1.1
* add compatibility with 12.X versions of Chef


test/integration/default/controls/default.rb
control 'elasticsearch_curator_default' do
  impact 1.0
  title 'Test configuration and install elasticsearch_curator'
  desc 'Test configuration and install elasticsearch_curator package method'

  describe package('elasticsearch-curator') do
    it { should be_installed }
  end

  describe file('/opt/elasticsearch-curator') do
    it { should be_directory }
  end

  describe file('/home/curator/.curator/curator.yml') do
    it { should be_file }
    its(:mode) { should cmp '0400' }
    it { should be_owned_by 'curator' }
  end

  # Test suites splitted between package install method and pip install method
  if os[:family] == 'debian'
    describe package('python-pkg-resources') do
      it { should be_installed }
    end
  # do nothing until I find how to test the APT repo
  elsif os[:family] == 'rhel'
    describe package('python-setuptools') do
      it { should be_installed }
    end
    describe file('/etc/yum.repos.d/elasticsearch-curator.repo') do
      it { should be_file }
    end
  end
end


test/integration/pip_install/controls/default.rb
control 'elasticsearch_curator_pip' do
  impact 1.0
  title 'Test configuration and install elasticsearch_curator'
  desc 'Test configuration and install elasticsearch_curator pip method'

  if os[:family] == 'debian'
    describe package('python-pip') do
      it { should be_installed }
    end
  elsif os[:family] == 'rhel'
    describe package('python2-pip') do
      it { should be_installed }
    end
    describe package('python-setuptools') do
      it { should be_installed }
    end
  end

  describe file('/home/curator/.curator/curator.yml') do
    it { should be_file }
    its(:mode) { should cmp '0400' }
    it { should be_owned_by 'curator' }
  end

  # Poise-python install virtualenv via pip, therefore resource package can't tell if it's install or not
  if os[:family] == 'debian'
    describe file('/usr/local/bin/virtualenv') do
      it { should be_file }
      it { should be_executable }
    end
  elsif os[:family] == 'rhel'
    describe file('/bin/virtualenv') do
      it { should be_file }
      it { should be_executable }
    end
  end

  describe command('python --version') do
    its(:exit_status) { should eq 0 }
  end

  # When installed via pip, package can't retrive the install

  describe file('/opt/elasticsearch-curator') do
    it { should be_directory }
  end

  describe file('/opt/elasticsearch-curator/bin/curator') do
    it { should be_file }
    its(:mode) { should cmp '0755' }
  end

  describe file('/opt/elasticsearch-curator/bin/curator_cli') do
    it { should be_file }
    its(:mode) { should cmp '0755' }
  end

  describe file('/opt/elasticsearch-curator') do
    it { should be_directory }
  end
end


test/integration/default/serverspec/default_spec.rb
# ServerSpec testing suited added by @GHostQC
require 'spec_helper'

describe package('elasticsearch-curator') do
  it { should be_installed }
end

describe file('/opt/elasticsearch-curator') do
  it { should be_directory }
end

# Test suites splitted between package install method and pip install method
if os[:family] == 'debian'
  describe package('python-pkg-resources') do
    it { should be_installed }
  end
# do nothing until I find how to test the APT repo
elsif os[:family] == 'rhel'
  describe package('python-setuptools') do
    it { should be_installed }
  end
  describe file('/etc/yum.repos.d/elasticsearch-curator.repo') do
    it { should be_file }
  end
end


resources/action.rb
# resources/action.rb
#
# Cookbook Name:: elasticsearch-curator
# Resource:: action
#
# Copyright 2016 Servers.com
#

property :config, Hash, default: {}
property :username, String, default: node['elasticsearch-curator']['username']
property :minute, String, default: node['elasticsearch-curator']['cron_minute']
property :hour, String, default: node['elasticsearch-curator']['cron_hour']
property :path, String, default: node['elasticsearch-curator']['action_file_path']
property :bin_path, String, default: node['elasticsearch-curator']['bin_path']

default_action :create

action :create do
  directory new_resource.path do
    recursive true
    action :create
  end

  require 'yaml'

  file "#{new_resource.path}/#{new_resource.name}.yml" do
    content YAML.dump(new_resource.config.to_hash)
    user new_resource.username
    mode '0644'
  end

  curator_args = "--config #{node['elasticsearch-curator']['config_file_path']}/curator.yml #{new_resource.path}/#{new_resource.name}.yml"
  cron_d "curator-#{new_resource.name}" do
    command "#{new_resource.bin_path}/curator #{curator_args}"
    user    new_resource.username
    minute  new_resource.minute
    hour    new_resource.hour
    action  [:create]
  end
end


resources/config.rb
# resources/config.rb
#
# Cookbook Name:: elasticsearch-curator
# Resource:: config
#
# Copyright 2016 Servers.com
#

property :config, Hash, default: {}
property :username, String, default: node['elasticsearch-curator']['username']
property :path, String, default: node['elasticsearch-curator']['config_file_path']
property :http_auth, [String, nil], default: nil
default_action :configure

action :configure do
  user new_resource.username do
    system true
    action :create
    manage_home true
  end

  group new_resource.username do
    members new_resource.username
    append true
    system true
  end

  directory new_resource.path do
    recursive true
    action :create
  end

  require 'yaml'

  curatorconfig = new_resource.config.to_hash.clone

  curatorconfig['client']['http_auth'] = new_resource.http_auth if new_resource.http_auth && new_resource.http_auth.length > 2 && new_resource.http_auth.include?(':')

  file "#{new_resource.path}/curator.yml" do
    content YAML.dump(curatorconfig.to_hash)
    user new_resource.username
    mode '0400'
    sensitive true
  end
end


resources/install.rb
# resources/install.rb
#
# Cookbook Name:: elasticsearch-curator
# Resource:: install
#
# Copyright 2016 Servers.com
#

property :arch_type, String
property :install_method, String, default: 'package'
property :repository_url, String, default: node['elasticsearch-curator']['repository_url']
property :repository_key, String, default: node['elasticsearch-curator']['repository_key']
property :install_version, [String, nil],  default: nil
default_action :install

action :install do
  if platform_family?('debian')
    apt_repository 'elasticsearch-curator' do
      uri new_resource.repository_url
      distribution ''
      components %w(stable main)
      key new_resource.repository_key
      only_if { new_resource.install_method == 'package' }
    end
  elsif platform_family?('rhel')
    yum_repository 'elasticsearch-curator' do
      baseurl new_resource.repository_url
      gpgkey new_resource.repository_key
      only_if { new_resource.install_method == 'package' }
    end
  else
    Chef::Log.warn "I do not support your platform: #{node['platform_family']}"
  end

  if new_resource.install_method == 'package'
    package 'elasticsearch-curator' do
      version node['elasticsearch-curator']['version']
    end
    if platform_family?('debian')
      package 'python-pkg-resources'
    elsif platform_family?('rhel')
      package 'python-setuptools'
    end
  elsif new_resource.install_method == 'pip'
    package 'python-pip'
    node.override['poise-python']['provider'] = 'system'
    node.override['poise-python']['provider'] = 'scl' if platform_family?('rhel')
    @run_context.include_recipe 'poise-python::default'
    python_virtualenv '/opt/elasticsearch-curator'
    python_package 'elasticsearch-curator' do
      version node['elasticsearch-curator']['version']
      action :install
    end
  else
    Chef::Application.fatal!("Unknown install method: #{@install_method}")
  end
end


attributes/default.rb
default['elasticsearch-curator']['version'] = nil
default['elasticsearch-curator']['install_method'] = 'package'
case node['platform_family']
when 'debian'
  default['elasticsearch-curator']['repository_url'] = 'http://packages.elastic.co/curator/5/debian'
when 'rhel'
  default['elasticsearch-curator']['repository_url'] = case node['platform_version'].to_i
                                                       when 6
                                                         'http://packages.elastic.co/curator/5/centos/6'
                                                       when 7
                                                         'http://packages.elastic.co/curator/5/centos/7'
                                                       end
end
default['elasticsearch-curator']['repository_key'] = 'https://packages.elastic.co/GPG-KEY-elasticsearch'
default['elasticsearch-curator']['bin_path'] = '/usr/bin'
default['elasticsearch-curator']['config'] = {
  'client' => {
    'hosts' => ['127.0.0.1'],
    'port' => 9200,
    'use_ssl' => false,
    'ssl_no_validate' => false,
    'timeout' => 30,
    'master_only' => false,
  },
  'logging' => {
    'loglevel' => 'INFO',
    'logformat' => 'default',
  },
}
default['elasticsearch-curator']['username'] = 'curator'
default['elasticsearch-curator']['config_file_path'] = "/home/#{node['elasticsearch-curator']['username']}/.curator"
default['elasticsearch-curator']['action_file_path'] = "/home/#{node['elasticsearch-curator']['username']}/.curator"
default['elasticsearch-curator']['cron_minute'] = '0'
default['elasticsearch-curator']['cron_hour'] = '*'
override['poise-python']['options']['wheel_version'] = '0.29.0' if node['platform_version'].to_i == 6
